### ISSUE TITLE
Bug: n/hashbang doesn't consider source file of a bin path as a binary

### ISSUE THREAD

**Issue by simonhaenisch** [eslint-community/eslint-plugin-n#304]
### Environment

Node version: 20
npm version: 🤷🏻‍♂️
ESLint version: 9
eslint-plugin-n version: latest
Operating System: macOS

### What rule do you want to report?

n/hashbang

### Link to Minimal Reproducible Example

should be clear enough? can provide later if necessary

### What did you expect to happen?

I have a file `src/cli.ts` that gets transpiled to `dist/cli.js`, so I have `"bin": "dist/cli.js"` in my `package.json`.

`n/hashbang` reports the shebang in the file as "This file needs no shebang" because I guess it doesn't know it's the file that gets transpiled into `dist/cli.js`.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I've tried looking at `convertPath` but don't see how that could solve it. I also saw there's some config related to tsconfig, and thinking based on `"outDir": "dist"` in the `tsconfig.json` it could actually be able to figure out automatically that `src/cli.ts` is the source file for `dist/cli.js`?

**Comment by aladdin-add** [comment]
`convertPath` should works:
```js
    settings: {
      node: {
        convertPath: [
          {
            include: ['src/*.ts'],
            replace: ['^src/(.+).ts$', 'dist/$1.js'],
          },
        ],
      },
    },
```

we can add an example to show how to use it in ts projects. :)

**Comment by simonhaenisch** [comment]
Can this convert path setting be added automatically based on `tsconfig.json`?

Just guessing but could be sth like

```js
const tsconfig = require('./tsconfig.json')

const { include, outDir } = tsconfig.compilerOptions ?? {};

if (!outDir) {
  return;
}

const convertPaths = include
  .filter(path => !path.includes('/') && !path.includes('*')) // only pick simple paths
  .map(path => ({
    include: [`${path/**/*.ts}`],
    replace: [`^${path}/(.+).ts$`, `${outDir}/$1.js`]
  })
```

Maybe it's too complex to solve like this though, but maintaining that kind of setting per project is quite a chore.

**Comment by scagood** [comment]
I suppose it'd be a reasonable complement to the import extension mapping :+1: 

https://github.com/eslint-community/eslint-plugin-n/blob/master/lib/util/get-typescript-extension-map.js#L21-L27
